models/currency.js

```diff
@@ -0,0 +1,188 @@
+    await currencyExchangeModel.doc(allCurrencyName).set({
+      [currencyData.src]: {
+        [currencyData.target]: currencyData.quantity
+      },
+      lastUpdate: Date.now()
+    }, {
+      merge: true
+    })
+    return true
+  } catch (err) {
+    logger.error('Error while adding currency rates into exchange', err)
+    throw err
+  }
+}
+
+/**
+ * Fetch the all banks availabe name and id
+ * @return {Promise<exchangeData>}
+ */
+const getAllBanks = async () => {
+  try {
+    const bankSnapshot = await bankModel.get()
+    const bankData = []
+    bankSnapshot.forEach(bank => {
+      const data = bank.data()
+      if (data.isActive) {
+        bankData.push({ bankId: data.bankId, bankName: data.bankName })
+      }
+    })
+    return bankData
+  } catch (err) {
+    logger.error('Error while getting bank detais', err)
+    throw err
+  }
+}
+
+/**
+ * Rounds number to 2 precision after decimal point
+ * @return {Promise<object>}
+ */
+const roundToTwoDigits = (num) => {
+  return Math.round(num * 100) / 100
+}
+
+/**
+ * Update bank currecy after excbange
+ * @return {Promise<exchangeData>}
+ */
+const updateBankWalletAfterExchange = (bankCurrency, exchangeData, totalTargetCurrencyRequest) => {
+  bankCurrency[exchangeData.src] = roundToTwoDigits((bankCurrency[exchangeData.src] || 0) + parseFloat(exchangeData.quantity))
+  bankCurrency[exchangeData.target] = roundToTwoDigits(bankCurrency[exchangeData.target] - totalTargetCurrencyRequest)
+  return { ...bankCurrency }
+}
+
+/**
+ * Update user currecy after excbange
+ * @return {Promise<exchangeData>}
+ */
+const updateUserWalletAfterExchange = (userCurrency, exchangeData, totalTargetCurrencyRequest) => {
+  userCurrency[exchangeData.target] = roundToTwoDigits((userCurrency[exchangeData.target] || 0) + parseFloat(totalTargetCurrencyRequest))
+  userCurrency[exchangeData.src] = roundToTwoDigits(userCurrency[exchangeData.src] - exchangeData.quantity)
+  return { ...userCurrency }
+}
+
+/**
+ * Transaction for currency exchange
+ * @return {Promise<exchangeData>}
+ */
+const exchangeTransaction = async (userId, exchangeData) => {
+  try {
+    const { bankId } = exchangeData
+    // db models refs
+    const userWalletRef = walletModel.where('userId', '==', userId).limit(1)
+    const bankWalletRef = bankModel.where('bankId', '==', bankId).limit(1)
+    const exchangeRateRef = currencyExchangeModel.doc(allCurrencyName)
+    const exchangeResponse = await firestore.runTransaction(async t => {
+      const responseObj = {
+        status: false,
+        message: ''
+      }
+
+      const userWalletDoc = await t.get(userWalletRef)
+      const userWalletId = await extractReferenceDocumentId(userWalletDoc)
+      const userWallet = await extractReferenceDocumentData(userWalletDoc)
+      const bankWalletDoc = await t.get(bankWalletRef)
+      const bankWalletId = await extractReferenceDocumentId(bankWalletDoc)
+      const bankWallet = await extractReferenceDocumentData(bankWalletDoc)
```

Comment:
Feel free to correct me here

Initally promise corresponding to `userWalletDoc` will resolve, then `userWalletId` then `userWallet` and then `bankWalletDoc` would get an opportunity to start executing.
`bankWalletDoc` has nothing to do with `userWalletDoc` or `userWalletId` or `userWallet`, right?
Aren't we then unnecessarily blocking the execution here?